### PR TITLE
Knowhere GPU support

### PR DESCRIPTION
**Pull request by Presburger**
Signed-off-by: yudong.cai <yudong.cai@zilliz.com>

### CODE REVIEW

File: knowhere/index/VecIndexFactory.cpp [hunk 1]
```diff
@@ -100,23 +100,23 @@ VecIndexFactory::CreateVecIndex(const IndexType& type, const IndexMode mode) {
...
         case IndexMode::MODE_GPU: {
-            auto gpu_device = -1;  // TODO: remove hardcode here, get from invoker
+            auto gpu_device = 0;  // TODO: remove hardcode here, get from invoker
             if (type == IndexEnum::INDEX_FAISS_BIN_IDMAP) {
                 return std::make_shared<knowhere::BinaryIDMAP>();
             } else if (type == IndexEnum::INDEX_FAISS_BIN_IVFFLAT) {
                 return std::make_shared<knowhere::BinaryIVF>();
             } else if (type == IndexEnum::INDEX_FAISS_IDMAP) {
-                return std::make_shared<knowhere::GPUIDMAP>(gpu_device);
+                return std::make_shared<knowhere::IDMAP>();
```
Comment (cydrain) [inline knowhere/index/VecIndexFactory.cpp:109]: wrong change

File: knowhere/index/VecIndexFactory.cpp [hunk 2]
```diff
@@ -100,23 +100,23 @@ VecIndexFactory::CreateVecIndex(const IndexType& type, const IndexMode mode) {
...
             } else if (type == IndexEnum::INDEX_FAISS_IVFFLAT) {
-                return std::make_shared<knowhere::GPUIVF_NM>(gpu_device);
+                return std::make_shared<knowhere::GPUIVF>(gpu_device);
```
Comment (cydrain) [inline knowhere/index/VecIndexFactory.cpp:111]: index with different index_mode should match each other, otherwise CopyCpuToGpu() and CopyGpuToCpu() cannot work.
By default we create IVF_NM for CPU flat, if change here to GPUIVF, they mis-match.

File: knowhere/index/vector_index/gpu/GPUIndex.h [hunk 3]
```diff
@@ -11,14 +11,16 @@
...
+#include "knowhere/index/VecIndex.h"
 #include "knowhere/index/vector_index/helpers/FaissGpuResourceMgr.h"
 
 namespace knowhere {
 
 class GPUIndex {
  public:
     explicit GPUIndex(const int& device_id) : gpu_id_(device_id) {
+        //TODO(yusheng.ma): param can set
+        knowhere::FaissGpuResourceMgr::GetInstance().InitDevice(device_id, 200 * 1024 * 1024, 300 * 1024 * 1024, 2);
```
Comment (cydrain) [inline knowhere/index/vector_index/gpu/GPUIndex.h:23]: InitDevice() should be called by invoker, not here.

File: knowhere/index/vector_index/gpu/IndexGPUIDMAP.h [hunk 4]
```diff
@@ -24,6 +24,10 @@ using Graph = std::vector<std::vector<int64_t>>;
 
 class GPUIDMAP : public IDMAP, public GPUIndex {
  public:
+    explicit GPUIDMAP(const int& device_id) : IDMAP(), GPUIndex(device_id) {
```
Comment (cydrain) [inline knowhere/index/vector_index/gpu/IndexGPUIDMAP.h:27]: why add this constructor ?

File: knowhere/index/vector_offset_index/IndexIVF_NM.cpp [hunk 5]
```diff
@@ -222,6 +222,12 @@ IVF_NM::CopyCpuToGpu(const int64_t device_id, const Config& config) {
 #ifdef KNOWHERE_GPU_VERSION
     if (auto res = FaissGpuResourceMgr::GetInstance().GetRes(device_id)) {
         ResScope rs(res, device_id, false);
+        /*
```
Comment (cydrain) [inline knowhere/index/vector_offset_index/IndexIVF_NM.cpp:225]: remove these lines

File: thirdparty/CMakeLists.txt [hunk 6]
```diff
@@ -71,47 +71,30 @@ macro(build_faiss)
...
+    endif()
+    set(FAISS_CMAKE_ARGS
             "-DCMAKE_INSTALL_PREFIX=${CMAKE_INSTALL_PREFIX}"
             "-DCMAKE_BUILD_TYPE=Release"
             "-DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}"
             "-DCMAKE_CUDA_COMPILER=${CMAKE_CUDA_COMPILER}"
 	    "-DCMAKE_CXX_FLAGS=${CMAKE_CXX_FLAGS} -fpermissive -I ${KNOWHERE_SOURCE_DIR} ${MMAN_INCLUDE_FLAGS}"
             "-DFAISS_ENABLE_PYTHON=OFF"
-            "-DFAISS_ENABLE_GPU=OFF"
+            "-DFAISS_ENABLE_GPU=${KNOWHERE_GPU_ENABLE}"
```
Comment (cydrain) [inline thirdparty/CMakeLists.txt:89]: need not add "KNOWHERE_GPU_ENABLE", use "KNOWHERE_GPU_VERSION" directly

File: thirdparty/faiss/CMakeLists.txt [hunk 7]
```diff
@@ -36,14 +36,11 @@ if (FAISS_ENABLE_CCACHE)
     endif (CCACHE_FOUND)
 endif()
 
+
```
Comment (cydrain) [inline thirdparty/faiss/CMakeLists.txt:39]: remove empty lines

File: thirdparty/faiss/CMakeLists.txt [hunk 8]
```diff
@@ -36,14 +36,11 @@ if (FAISS_ENABLE_CCACHE)
...
 if(FAISS_ENABLE_GPU)
-  # externalproject_add use semicolon to separate commands, revert to desired format
-  if (CMAKE_CUDA_ARCHITECTURES)
-    string(REPLACE ":" ";" CMAKE_CUDA_ARCHITECTURES "${CMAKE_CUDA_ARCHITECTURES}")
-  endif()
-  set(CMAKE_CUDA_HOST_COMPILER ${CMAKE_CXX_COMPILER})
   enable_language(CUDA)
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DUSE_GPU")
+  set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -I${FAISS_INCLUDE_DIR} -fPIC")
```
Comment (cydrain) [inline thirdparty/faiss/CMakeLists.txt:43]: CMAKE_CUDA_FLAGS should be passed in from knowhere

File: thirdparty/faiss/faiss/gpu/GpuIndexFlat.cu [hunk 9]
```diff
@@ -126,7 +126,8 @@ void GpuIndexFlat::copyFrom(const faiss::IndexFlat* index) {
     xb_.clear();
 
     if (flatConfig_.storeInCpu) {
-        xb_ = index->xb;
+        xb_.resize(index->codes.size()/4);
```
Comment (cydrain) [inline thirdparty/faiss/faiss/gpu/GpuIndexFlat.cu:129]: wrong change

File: thirdparty/faiss/faiss/gpu/test/CMakeLists.txt [hunk 10]
```diff
@@ -33,4 +34,4 @@ add_executable(demo_ivfpq_indexing_gpu EXCLUDE_FROM_ALL
   demo_ivfpq_indexing_gpu.cpp)
 
 target_link_libraries(demo_ivfpq_indexing_gpu
-  PRIVATE faiss gtest_main CUDA::cudart)
+  PRIVATE  faiss gtest_main CUDA::cudart)
```
Comment (cydrain) [inline thirdparty/faiss/faiss/gpu/test/CMakeLists.txt:37]: remove unnecessary change

File: unittest/test_gpuresource.cpp [hunk 11]
```diff
@@ -114,36 +115,37 @@ TEST_F(GPURESTEST, copyandsearch) {
...
-    auto search_idx = milvus::knowhere::cloner::CopyCpuToGpu(cpu_idx, DEVICEID, milvus::knowhere::Config());
+    auto search_idx = knowhere::cloner::CopyCpuToGpu(cpu_idx, DEVICE_ID, knowhere::Config());
 
     constexpr int train_count = 5;
     constexpr int search_count = 200;
     auto train_stage = [&] {
         for (int i = 0; i < train_count; ++i) {
-            index_->Train(base_dataset, conf);
-            index_->AddWithoutIds(base_dataset, conf);
+            search_idx->Train(base_dataset, conf);
```
Comment (cydrain) [inline unittest/test_gpuresource.cpp:141]: should not change

File: unittest/test_gpuresource.cpp [hunk 12]
```diff
@@ -114,36 +115,37 @@ TEST_F(GPURESTEST, copyandsearch) {
...
+            search_idx->AddWithoutIds(base_dataset, conf);
         }
     };
-    auto search_stage = [&](milvus::knowhere::VecIndexPtr& search_idx) {
+    auto search_stage = [&](knowhere::VecIndexPtr& search_idx) {
         for (int i = 0; i < search_count; ++i) {
             auto result = search_idx->Query(query_dataset, conf, nullptr);
+            std::cout<<conf<<std::endl;
```
Comment (cydrain) [inline unittest/test_gpuresource.cpp:148]: remove this line

File: unittest/test_idmap.cpp [hunk 13]
```diff
@@ -96,13 +97,14 @@ TEST_P(IDMAPTest, idmap_basic) {
     auto result = index_->Query(query_dataset, conf, nullptr);
     AssertAnns(result, nq, k);
     // PrintResult(result, nq, k);
-
+/*
```
Comment (cydrain) [inline unittest/test_idmap.cpp:100]: should not comment this GPU check

File: unittest/test_ivf.cpp [hunk 14]
```diff
@@ -74,7 +75,7 @@ INSTANTIATE_TEST_CASE_P(
     IVFParameters,
     IVFTest,
     Values(
-#ifdef KNOWHERE_GPU_VERSION
+#if 0
         std::make_tuple(knowhere::IndexEnum::INDEX_FAISS_IVFPQ, knowhere::IndexMode::MODE_GPU),
         std::make_tuple(knowhere::IndexEnum::INDEX_FAISS_IVFSQ8, knowhere::IndexMode::MODE_GPU),
         std::make_tuple(knowhere::IndexEnum::INDEX_FAISS_IVFSQ8H, knowhere::IndexMode::MODE_GPU),
```
Comment (cydrain) [inline unittest/test_ivf.cpp:81]: except IVFSQ8H, other GPU checks should pass

File: unittest/test_ivf.cpp [hunk 15]
```diff
@@ -264,6 +265,7 @@ TEST_P(IVFTest, clone_test) {
 }
 #endif
 
+#if 0
```
Comment (cydrain) [inline unittest/test_ivf.cpp:268]: should not comment this GPU check

File: unittest/test_ivf_nm.cpp [hunk 16]
```diff
@@ -55,7 +55,7 @@ class IVFNMTest : public DataGen,
 
     void
     TearDown() override {
-#ifdef KNOWHERE_GPU_VERSION
+#if 0
```
Comment (cydrain) [inline unittest/test_ivf_nm.cpp:58]: don't disable this

File: unittest/test_ivf_nm.cpp [hunk 17]
```diff
@@ -109,25 +109,36 @@ TEST_P(IVFNMTest, ivfnm_basic) {
     auto result = index_->Query(query_dataset, conf_, nullptr);
     AssertAnns(result, nq, k);
 
-#ifdef KNOWHERE_GPU_VERSION
+#if 0
```
Comment (cydrain) [inline unittest/test_ivf_nm.cpp:112]: don't disable this

File: unittest/test_ivf_nm.cpp [hunk 18]
```diff
@@ -109,25 +109,36 @@ TEST_P(IVFNMTest, ivfnm_basic) {
...
         EXPECT_NO_THROW({
-            auto clone_index = knowhere::cloner::CopyCpuToGpu(index_, DEVICEID, conf_);
+            auto clone_index = knowhere::cloner::CopyCpuToGpu(index_, DEVICE_ID, conf_);
             auto clone_result = clone_index->Query(query_dataset, conf_, nullptr);
             AssertAnns(clone_result, nq, k);
             std::cout << "clone C <=> G [" << index_type_ << "] success" << std::endl;
         });
     }
 
+    auto AssertEqual = [&](knowhere::DatasetPtr p1, knowhere::DatasetPtr p2) {
```
Comment (cydrain) [inline unittest/test_ivf_nm.cpp:124]: remove these lines

File: thirdparty/faiss/CMakeLists.txt (hunk 8 above)
Comment (Presburger) [reply thirdparty/faiss/CMakeLists.txt:43]: CMAKE_CUDA_FLAGS is cmake buildin  variables.

Comment (Presburger) [reply thirdparty/faiss/CMakeLists.txt:43]: > CMAKE_CUDA_FLAGS should be passed in from knowhere

CMAKE_CUDA_FLAGS is cmake buildin variables.

File: unittest/test_customized_index.cpp [hunk 19]
```diff
@@ -14,8 +14,10 @@
 
 #include "knowhere/common/Timer.h"
 #include "knowhere/index/IndexType.h"
+#include "knowhere/index/VecIndexFactory.h"
 #include "unittest/Helper.h"
 #include "unittest/utils.h"
+using json = nlohmann::json;
```
Comment (cydrain) [inline unittest/test_customized_index.cpp:20]: remove this line, use knowhere/common/Config.h

File: thirdparty/faiss/CMakeLists.txt [hunk 20]
```diff
@@ -37,13 +37,8 @@ if (FAISS_ENABLE_CCACHE)
...
 
 if(FAISS_ENABLE_GPU)
-  # externalproject_add use semicolon to separate commands, revert to desired format
-  if (CMAKE_CUDA_ARCHITECTURES)
-    string(REPLACE ":" ";" CMAKE_CUDA_ARCHITECTURES "${CMAKE_CUDA_ARCHITECTURES}")
-  endif()
-  set(CMAKE_CUDA_HOST_COMPILER ${CMAKE_CXX_COMPILER})
   enable_language(CUDA)
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DUSE_GPU")
+  set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -I${FAISS_INCLUDE_DIR} -fPIC")
```
Comment (cydrain) [inline thirdparty/faiss/CMakeLists.txt:41]: don't add "-fPIC" here, pass following config from knowhere:
"-DCMAKE_CUDA_FLAGS=-Xcompiler -fPIC -std=c++11 -D_FORCE_INLINES --expt-extended-lambda"

File: knowhere/index/vector_offset_index/gpu/IndexGPUIVF_NM.cpp [hunk 21]
```diff
@@ -61,11 +61,6 @@ GPUIVF_NM::AddWithoutIds(const DatasetPtr& dataset_ptr, const Config& config) {
     }
 }
 
-void
```
Comment (cydrain) [inline knowhere/index/vector_offset_index/gpu/IndexGPUIVF_NM.cpp:64]: don't remove this

File: knowhere/index/vector_offset_index/gpu/IndexGPUIVF_NM.cpp [hunk 22]
```diff
@@ -117,14 +112,37 @@ GPUIVF_NM::SerializeImpl(const IndexType& type) {
         std::shared_ptr<uint8_t[]> data(writer.data_);
 
         BinarySet res_set;
-        res_set.Append("IVF", data, writer.rp);
+        res_set.Append("IVF_NM", data, writer.rp);
```
Comment (cydrain) [inline knowhere/index/vector_offset_index/gpu/IndexGPUIVF_NM.cpp:115]: don't change this

File: thirdparty/faiss/CMakeLists.txt [hunk 23]
```diff
@@ -37,13 +37,7 @@ if (FAISS_ENABLE_CCACHE)
 endif()
 
 if(FAISS_ENABLE_GPU)
-  # externalproject_add use semicolon to separate commands, revert to desired format
-  if (CMAKE_CUDA_ARCHITECTURES)
-    string(REPLACE ":" ";" CMAKE_CUDA_ARCHITECTURES "${CMAKE_CUDA_ARCHITECTURES}")
-  endif()
-  set(CMAKE_CUDA_HOST_COMPILER ${CMAKE_CXX_COMPILER})
   enable_language(CUDA)
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DUSE_GPU")
```
Comment (cydrain) [inline thirdparty/faiss/CMakeLists.txt:46]: this line is necessary

File: thirdparty/faiss/faiss/gpu/GpuIndexFlat.cu [hunk 24]
```diff
@@ -126,7 +126,8 @@ void GpuIndexFlat::copyFrom(const faiss::IndexFlat* index) {
     xb_.clear();
 
     if (flatConfig_.storeInCpu) {
-        xb_ = index->xb;
+        xb_.resize(index->codes.size()/sizeof(float));
+        memcpy(&xb_[0], &index->codes[0], index->codes.size());
```
Comment (cydrain) [inline thirdparty/faiss/faiss/gpu/GpuIndexFlat.cu:130]: don't use "index->codes[0]", use "index->get_xb()"

File: thirdparty/faiss/CMakeLists.txt [hunk 25]
```diff
@@ -37,13 +37,9 @@ if (FAISS_ENABLE_CCACHE)
 endif()
 
 if(FAISS_ENABLE_GPU)
-  # externalproject_add use semicolon to separate commands, revert to desired format
-  if (CMAKE_CUDA_ARCHITECTURES)
-    string(REPLACE ":" ";" CMAKE_CUDA_ARCHITECTURES "${CMAKE_CUDA_ARCHITECTURES}")
-  endif()
   set(CMAKE_CUDA_HOST_COMPILER ${CMAKE_CXX_COMPILER})
-  enable_language(CUDA)
```
Comment (cydrain) [inline thirdparty/faiss/CMakeLists.txt:45]: should keep this line

File: unittest/test_idmap.cpp [hunk 26]
```diff
@@ -286,13 +288,13 @@ TEST_P(IDMAPTest, idmap_copy) {
...
     ASSERT_ANY_THROW(knowhere::cloner::CopyCpuToGpu(index_, -1, conf));
-    auto clone_index = knowhere::cloner::CopyCpuToGpu(index_, DEVICEID, conf);
+    auto clone_index = knowhere::cloner::CopyCpuToGpu(index_, DEVICE_ID, conf);
     auto clone_result = clone_index->Query(query_dataset, conf, nullptr);
 
     AssertAnns(clone_result, nq, k);
     ASSERT_THROW({ std::static_pointer_cast<knowhere::GPUIDMAP>(clone_index)->GetRawVectors(); },
                  knowhere::KnowhereException);
-    ASSERT_ANY_THROW(clone_index->Serialize(conf));
+    ASSERT_NO_THROW(clone_index->Serialize(conf));
```
Comment (cydrain) [inline unittest/test_idmap.cpp:297]: you can comment this line, don't change it.
We expect some exception throw out, if it does not throw exception any more, we need check what's changed.
